pkg/admission/prep_downscale.go

```diff
@@ -0,0 +1,178 @@
+	// while for a Scale object, it will be the labels of the StatefulSet/Deployment/ReplicaSet that the Scale object belongs to.
+	var lbls map[string]string
+	switch o := oldObj.(type) {
+	case *appsv1.Deployment:
+		lbls = o.Labels
+	case *appsv1.StatefulSet:
+		lbls = o.Labels
+	case *appsv1.ReplicaSet:
+		lbls = o.Labels
+	case *autoscalingv1.Scale:
+		lbls, err = getResourceLabels(ctx, ar, api)
+		if err != nil {
+			return allowBecauseCannotGetResource(ar, logger, err)
+		}
+	default:
+		return allowWarn(logger, fmt.Sprintf("unsupported type %T, allowing the change", o))
+	}
+
+	// Since it's a downscale, check if the resource has the label that indicates it's ready to be prepared to be downscaled.
+	// Create a slice of endpoint addresses for pods to send HTTP post requests to and to fail if any don't return 200
+	if !*ar.Request.DryRun && lbls[PrepDownscaleLabelKey] == PrepDownscaleLabelValue {
+		diff := (*oldReplicas - *newReplicas)
+		type endpoint struct {
+			url   string
+			index int
+		}
+		eps := make([]endpoint, diff)
+
+		// The DNS entry for a pod of a stateful set is
+		// ingester-zone-a-0.$(servicename).$(namespace).svc.cluster.local
+		// The service in this case is ingester-zone-a as well.
+		// https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/#stable-network-id
+		for i := 0; i < int(diff); i++ {
+			eps[i].url = fmt.Sprintf("%v-%v.%v.%v.svc.cluster.local:%s/ingester/%s",
+				ar.Request.Name,       // pod name
+				int(*oldReplicas)-i-1, // nr in statefulset
+				ar.Request.Name,       // svc name
+				ar.Request.Namespace,
+				lbls[PrepDownscalePortKey],
+				lbls[PrepDownscalePathKey],
+			)
+			eps[i].index = int(*oldReplicas) - i - 1
+		}
+
+		g, _ := errgroup.WithContext(ctx)
+		for _, ep := range eps {
+			ep := ep // https://golang.org/doc/faq#closures_and_goroutines
+			g.Go(func() error {
+				resp, err := client.Post("http://"+ep.url, "application/json", nil)
+				if err != nil {
+					level.Error(logger).Log("msg", "error sending HTTP post request", "endpoint", ep.url, "err", err)
+					return err
+				}
+				if resp.StatusCode != 200 {
+					err := fmt.Errorf("HTTP post request returned non-200 status code")
+					level.Error(logger).Log("msg", err, "endpoint", ep.url, "status", resp.StatusCode)
+					return err
+				}
+				if resp.StatusCode == 200 {
+					// TODO(jordanrushing): set a label on the pod to indicate it's been prepared for shutdown? Or do we need to do that?
+					// If we did that, when the rollout operator reconcile loop runs, it would need to "de-prep" the pod and remove the label
+					// OR
+					// we could do that in the conditional below
+					level.Debug(logger).Log("msg", "pod prepared for shutdown", "endpoint", ep.url)
+
+					err = addDownscaledAnnotationToPod(ctx, api, ar.Request.Namespace, ar.Request.Name, ep.index)
+					level.Debug(logger).Log("msg", "annotation added to pod", "err", err)
```

Comment:
Log which pod, please, and change the message if `err != nil`

```suggestion
					level.Debug(logger).Log("msg", "annotation added to pod", "err", err, "url", ep.url)
```